### PR TITLE
[SPARK-14536] [SQL] fix to handle null value in array type column for postgres.

### PR DESCRIPTION
**Pull request by sureshthalamati**
## What changes were proposed in this pull request?

JDBC  read  is failing with  NPE due to missing null value check for array data type if the source table has null values in the array type column.  For null values Resultset.getArray()  returns null. 
This PR adds null safe check to the Resultset.getArray() value before invoking method on the Array object. 
## How was this patch tested?

Updated the PostgresIntegration test suite to test null values. Ran docker integration tests on my laptop.

### CODE REVIEW

**Comment by sureshthalamati** [comment]
@JoshRosen

**Comment by sureshthalamati** [comment]
@srowen SPARK-14536 is not a duplicate, this PR will  address the issue.

**Comment by miketrewartha** [comment]
@sureshthalamati Any plans to get the conflicts fixed here and get things merged?

**Comment by gatorsmile** [comment]
The original support by array type is done in https://github.com/apache/spark/pull/9662. Could you resolve the conflicts and reproduce the error in the latest code base?

**Comment by sureshthalamati** [comment]
sure. I will  resolve the conflicts today.

**Comment by gatorsmile** [comment]
Thanks!

**Comment by sureshthalamati** [comment]
@gatorsmile Verified on the master , problem still exist.  Resolved the conflicts, when u get a chance can you please review.

Error stack without the fix 
java.lang.NullPointerException
	at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$$anonfun$org$apache$spark$sql$execution$datasources$jdbc$JdbcUtils$$makeGetter$13.apply(JdbcUtils.scala:469)
	at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$$anonfun$org$apache$spark$sql$execution$datasources$jdbc$JdbcUtils$$makeGetter$13.apply(JdbcUtils.scala:467)
	at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$$anon$1.getNext(JdbcUtils.scala:328)
	at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$$anon$1.getNext(JdbcUtils.scala:310)
	at org.apache.spark.util.NextIterator.hasNext(NextIterator.scala:73)
	at org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:39)
	at org.apache.spark.util.CompletionIterator.hasNext(CompletionIterator.scala:32)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIterator.processNext(Unknown Source)
	at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
	at org.apache.spark.sql.execution.WholeStageCodegenExec$$anonfun$8$$anon$1.hasNext(WholeStageCodegenExec.scala:377)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$2.apply(SparkPlan.scala:231)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$2.apply(SparkPlan.scala:225)
.....

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 1]
```diff
@@ -465,9 +465,9 @@ object JdbcUtils extends Logging {
       }
 
       (rs: ResultSet, row: InternalRow, pos: Int) =>
-        val array = nullSafeConvert[Object](
-          rs.getArray(pos + 1).getArray,
-          array => new GenericArrayData(elementConversion.apply(array)))
+        val array = nullSafeConvert[java.sql.Array](
+          rs.getArray(pos + 1),
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: Nit: `input = rs.getArray(pos + 1),`

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/PostgresIntegrationSuite.scala [hunk 2]
```diff
@@ -96,6 +100,9 @@ class PostgresIntegrationSuite extends DockerJDBCIntegrationSuite {
     assert(rows(0).getString(14) == "d1")
     assert(rows(0).getFloat(15) == 1.01f)
     assert(rows(0).getShort(16) == 1)
+
+    // Test reading null values.
```
Comment (gatorsmile) [inline external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/PostgresIntegrationSuite.scala]: // Test reading null values using the second row.

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/PostgresIntegrationSuite.scala [hunk 3]
```diff
@@ -51,12 +51,16 @@ class PostgresIntegrationSuite extends DockerJDBCIntegrationSuite {
...
+    ).executeUpdate()
   }
 
   test("Type mapping for various types") {
     val df = sqlContext.read.jdbc(jdbcUrl, "bar", new Properties)
-    val rows = df.collect()
-    assert(rows.length == 1)
+    val rows = df.collect().sortBy(_.toString())
+    assert(rows.length == 2)
     val types = rows(0).toSeq.map(x => x.getClass)
```
Comment (gatorsmile) [inline external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/PostgresIntegrationSuite.scala]: Add a comment above this line to indicate the following statements are testing the first row.

**Comment by gatorsmile** [comment]
Will also run the docker integration tests in my local computer. Post the results later. Thanks!

**Comment by gatorsmile** [comment]
LGTM pending test. The docker integration test cases pass in my local computer. 

However, I saw a not-related test case failure:
```
===== FINISHED o.a.s.sql.jdbc.OracleIntegrationSuite: 'SPARK-16625: General data types to be mapped to Oracle' =====

- SPARK-16625: General data types to be mapped to Oracle *** FAILED ***
  types.apply(9).equals("class java.sql.Date") was false (OracleIntegrationSuite.scala:136)
```

Submitted a JIRA: https://issues.apache.org/jira/browse/SPARK-19318

@sureshthalamati could you take a look at it and fix it? Thanks!

**Comment by gatorsmile** [comment]
Thanks! Merging to master.

**Comment by sureshthalamati** [comment]
Thank you, @gatorsmile

**Comment by miketrewartha** [comment]
Thanks guys, this will be super, super helpful to us!

**Comment by holdenk** [comment]
Hi @gatorsmile , on the dev@ list someone asked if this was something we were considering backporting into the 2.1 branch (for the 2.1.1 release). It looks like it might be a reasonable candidate for this so I figured since you were the one who committed it you might want to take a look and decide?

**Comment by gatorsmile** [comment]
@holdenk Sure, we can backport it to Spark 2.1. 

@sureshthalamati Could you please back port it to Spark 2.1?

**Comment by sureshthalamati** [comment]
sure. Thanks

**Comment by sureshthalamati** [comment]
Created PR to  back port this fix  to 2.1: https://github.com/apache/spark/pull/17460

**Pull request closed without merging**